### DISCUSSION TITLE
Filter count

### DISCUSSION THREAD

**Discussion by krazyjakee** [directus/directus#12853] [Archive]
@rijkvanzanten @azrikahar 
This is continuing the discussion on this pull request: https://github.com/directus/directus/pull/11737

The count filter `?filter[count(your_m2m_field)][_eq]=0` is very intensive on the database, even if it has a few thousand records.

root: 7000 records with
your_m2m_field: 3000 records

The database vendor is postgres at render.com.

```
RAM: 1 GB
CPU: 1
STORAGE: 16 GB
```

The request takes up to 15 to 30 seconds

Does this seem right?

**Comment by rijkvanzanten** [comment]
It's a fairly heavy query to perform, as the database effectively has to lookup the COUNT() of a subquery per parent row in order to run the filter 🤔 That being said, 15-30 seconds feels longer than expected still

**Reply by krazyjakee** [reply]
Thanks for the reply.

If there is no work around within Directus, I could cache the values for my use case but to really solve the problem, perhaps Directus could implement counter caches at the ORM level? That would make a broad number of use cases much faster.

**Reply by krazyjakee** [reply]
Here is a sequelize hook https://gist.githubusercontent.com/jordanell/e7f6916d8c9a92a1d1a0455ac7f8c5e6/raw/7b74fe35203b579569971015eb23fbc488c661b8/counterCache.js

**Comment by bernatvadell** [comment]
Perhaps the exists & not exists functionality could be implemented, something similar to:

`filter[(n?)exists(m2m_field)][nested_field][_eq]=bla`

expanded obj:
```json
{
  "_exists": {
    "m2m_field": {
      "_eq":  "bla"
    }
  }
}
```

in this example and the previous one, the filter of the relation itself would be included automatically.
```json
{
  "_exists": {
    "m2m_field": true
  }
}
```

In the fix I made to solve task #5996 I already tried the exists & not exists implementation:
![image](https://user-images.githubusercontent.com/17298604/164167555-f83ac322-72f9-40a8-b216-d69948f55f2a.png)

**Reply by rijkvanzanten** [reply]
> Perhaps the exists & not exists functionality could be implemented, something similar to:


That actually exists currently! You can use `_some` and `_none` to negate the o2m filter lookup 🙂

**Reply by bernatvadell** [reply]
> > Perhaps the exists & not exists functionality could be implemented, something similar to:
> 
> That actually exists currently! You can use `_some` and `_none` to negate the o2m filter lookup 🙂

true, would it be possible to use it in the following way?

`?filter[_none(your_m2m_field)]`

Indirectly, the relation m2m or o2m would apply.

I think it would be a much more optimal version than performing a subquery with a count.

**Reply by rijkvanzanten** [reply]
It's currently done like `?filter[categories][_none][name][_eq]=Recipes` 👍🏻

**Reply by bernatvadell** [reply]
what I am looking for is to filter by all the products that do not have any category

Similar to raw SQL:
```sql
select * from products where not exists (select * from product_category where product_category.product_id = products.id)
```

**Reply by krazyjakee** [reply]
I fiddled around and found this:
`?fields=id&filter[product_category][_none][id][_gt]=0`
That produced the result I expected much like your sequel query.

**Reply by bernatvadell** [reply]
You're asumming that the Id is a number.

**Reply by krazyjakee** [reply]
I'm assuming nothing about that last section after [_none], it's literally guess work as @rijkvanzanten and the docs didn't explain what it was actually doing.
